source/common/network/filter_matcher.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "envoy/config/listener/v3/listener_components.pb.h"
+#include "envoy/network/filter.h"
+#include "envoy/network/listen_socket.h"
+
+#include "absl/container/fixed_array.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * The helper to transform ListenerFilterChainMatchPredicate message to single matcher.
+ */
+class ListenerFilterMatcherBuilder {
+public:
+  static ListenerFilterMatcherPtr buildListenerFilterMatcher(
+      const envoy::config::listener::v3::ListenerFilterChainMatchPredicate& match_config);
+};
+
+/**
+ * Any matcher (always matches).
+ */
+class ListenerFilterAnyMatcher final : public ListenerFilterMatcher {
+public:
+  bool matches(ListenerFilterCallbacks&) const override { return true; }
+};
+
+class ListenerFilterNotMatcher final : public ListenerFilterMatcher {
+public:
+  ListenerFilterNotMatcher(
+      const envoy::config::listener::v3::ListenerFilterChainMatchPredicate& match_config)
+      : sub_matcher_(ListenerFilterMatcherBuilder::buildListenerFilterMatcher(match_config)) {}
+  bool matches(ListenerFilterCallbacks& cb) const override { return !sub_matcher_->matches(cb); }
+
+private:
+  const ListenerFilterMatcherPtr sub_matcher_;
+};
+
+/**
+ * Destination port matcher.
+ */
+class ListenerFilterDstPortMatcher final : public ListenerFilterMatcher {
+public:
+  explicit ListenerFilterDstPortMatcher(const ::envoy::type::v3::Int32Range& range)
+      : start_(range.start()), end_(range.end()) {}
+  bool matches(ListenerFilterCallbacks& cb) const override {
+    const auto& address = cb.socket().localAddress();
+    // Match on destination port (only for IP addresses).
+    if (address->type() == Address::Type::Ip) {
+      const auto port = address->ip()->port();
+      return start_ <= port && port < end_;
+    } else {
+      return true;
+    }
+  }
+
+private:
+  const uint32_t start_;
+  const uint32_t end_;
+};
+
+/**
+ * Matcher for implementing set logic.
+ */
+class ListenerFilterSetLogicMatcher : public ListenerFilterMatcher {
+public:
+  explicit ListenerFilterSetLogicMatcher(
+      absl::Span<const ::envoy::config::listener::v3::ListenerFilterChainMatchPredicate* const>
+          predicates);
+
+protected:
+  absl::FixedArray<ListenerFilterMatcherPtr> sub_matchers_;
```

Comment:
I think you probably want InlineArray here. AFAIK FixedArray is typically used for local stack allocations.

Reply:
Sorry I don't know which InlineArray you are referring to.
FixedArray beats vector because FixedArray does inline storage when # of element is low. And that is why we use it on stack.
I bet you are not referring to std::array because it require # of element at compile time

Reply:
Sorry I meant https://github.com/abseil/abseil-cpp/blob/master/absl/container/inlined_vector.h.

/wait